in_toto/artifact_rules.py

```diff
@@ -0,0 +1,186 @@
+  """
+  Parses the rule and extracts and returns the necessary data to apply the
+  rule. Can also be used to verify if a rule complies with any of the formats
+
+  <Arguments>
+    rule:
+        The rule to be unpacked, one of:
+        MATCH <pattern> [IN <source-path-prefix>] WITH (MATERIALS|PRODUCTS)
+            [IN <destination-path-prefix>] FROM <step>,
+        CREATE <path,
+        DELETE <pattern>,
+        MODIFY <pattern>,
+        ALLOW <pattern>,
+        DISALLOW <pattern>
+
+  <Exceptions>
+    raises FormatError, if the rule does not comply with any of the formats.
+
+  <Side Effects>
+    None.
+
+  <Returns>
+    A dictionary of the artifact rule data,
+    if it is a generic rule the dictionary is:
+    {
+      "type": rule[0], i.e.: ("CREATE"|"MODIFY"|"DELETE"|"ALLOW"|"DISALLOW")
+      "pattern" : rule[1], i.e. path pattern
+    }
+
+    if it is a match rule, the dictionary is:
+    {
+      "type": rule[0], i.e. "MATCH",
+      "pattern": rule[1], i.e. path pattern
+      "source_prefix": path or empty string
+      "dest_prefix": path or empty string
+      "dest_type" : destination artifact type, ("MATERIAL"|"PRODUCT")
+      "dest_name": destination step/inspection name
+    }
+
+  """
+
+  GENERIC_RULES = {"create", "modify", "delete", "allow", "disallow",}
+  COMPLEX_RULES = {"match",}
+  ALL_RULES = GENERIC_RULES | COMPLEX_RULES
+
+  if not isinstance(rule, list):
+    raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+        " rule must be a list.\n"
+        "Got: \n\t'{0}'".format(rule))
+
+  rule_len = len(rule)
+
+  if (rule_len < 2 or not isinstance(rule[0], six.string_types) or
+      rule[0].lower() not in ALL_RULES):
+    raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+        " rule must start with one of '{0}' and specify a 'pattern' as"
+        " second element.\n"
+        "Got: \n\t'{1}'".format(", ".join(ALL_RULES), rule))
+
+  rule_type = rule[0].lower()
+  pattern = rule[1]
+
+  if not isinstance(pattern, six.string_types):
+    raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+        " 'pattern' (second element) must be a string.\n"
+        "Got: \n\t'{0}'".format(rule))
+
+  # Type is one of "CREATE", "MODIFY", "DELETE", "ALLOW", "DISALLOW"
+  if rule_type in GENERIC_RULES:
+    if rule_len != 2:
+      raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+        " generic rules must have one of the formats:\n\t"
+        "CREATE <pattern>\n\t"
+        "MODIFY <pattern>\n\t"
+        "DELETE <pattern>\n\t"
+        "ALLOW <pattern>\n\t"
+        "DISALLOW <pattern>\n"
+        "Got:\n\t{}".format(rule))
+    else:
+      return {
+        "type": rule_type,
+        "pattern": pattern,
+      }
+
+  # Type is "MATCH"
+  elif rule_type in COMPLEX_RULES:
+
+    # ... IN <source-path-prefix> WITH (MATERIALS|PRODUCTS)
+    # IN <destination-path-prefix> FROM <step>
+    if (rule_len == 10 and rule[2].lower() == "in" and
```

Comment:
I wonder if it would improve readabilty to lower() all the strings before starting comparisons. e.g.,
```
rule = [x.lower for x in rule]
```